py/obj.c

```diff
@@ -205,21 +202,41 @@ bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
             mp_warning(MP_WARN_CAT(BytesWarning), "Comparison between bytes and str");
         }
         #endif
-        return false;
+        return local_false;
     }
 
     // generic type, call binary_op(MP_BINARY_OP_EQUAL)
-    mp_obj_type_t *type = mp_obj_get_type(o1);
-    if (type->binary_op != NULL) {
-        mp_obj_t r = type->binary_op(MP_BINARY_OP_EQUAL, o1, o2);
-        if (r != MP_OBJ_NULL) {
-            return r == mp_const_true ? true : false;
+    for (int pass = 0; pass < 2; pass++) {
+        mp_obj_type_t *type = mp_obj_get_type(o1);
+        if (type->binary_op != NULL) {
+            mp_binary_op_t op = not_equal ? MP_BINARY_OP_NOT_EQUAL : MP_BINARY_OP_EQUAL;
+            mp_obj_t r = type->binary_op(op, o1, o2);
+            if (r != MP_OBJ_NULL) {
+                return r;
+            }
+
+            // CPython is asymetric; it will try __eq__ if there is no
+            // __ne__ but not the other way around
+            if (not_equal) {
+                r = type->binary_op(MP_BINARY_OP_EQUAL, o1, o2);
+                if (r != MP_OBJ_NULL) {
+                    return mp_obj_is_true(r) ? mp_const_false : mp_const_true;
+                }
+            }
         }
+
+        // Try the other way around if none of the above worked
+        mp_obj_t temp = o1;
+        o1 = o2;
+        o2 = temp;
     }
 
-    // equality not implemented, and objects are not the same object, so
-    // they are defined as not equal
-    return false;
+    // equality not implemented, so fall back to pointer conparison
+    return (o1 == o2) ? local_true : local_false;
+}
+
+bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
+    return mp_obj_is_true(mp_obj_equal_bop(o1, o2, false));
```

Comment:
This is the function that matters for performance, because it's used in things like dict lookup and sequence comparison.  Prior to the change here it was "fast" because it was just returning true/false (C values) directly.  But now it must go through `mp_obj_is_true()` to convert the object to a C-level boolean.  Of course this is more powerful (and correct wrt to CPython).  To get back prior performance (if there was any) probably need to inline part of `mp_obj_is_true` here (eg the test for `mp_const_true`/`mp_const_false`).  I'm not saying this needs to be done, just considering the impact.

Reply:
Originally I implemented this in line and then reverted to this version because I was concerned about code size. I wonder if it might make sense to set up some sort of conditionally in-lining based on a setting on port configuration.